File: src/server/src/velvet_dawn/mechanics/triggers.py
from typing import Dict, List

import velvet_dawn
from velvet_dawn import errors
from velvet_dawn.db.instances import Instance
from velvet_dawn.mechanics.actions.action import Action


""" Triggers class attached to units and tiles. This
class also parses the given dict of triggers and 
actions given to the class
"""


# TODO implement attack and attacked, spawn, death and kill
# TODO Add validation for this
WORLD_TRIGGERS = ["game", "turn", "turn-end"]
TILE_TRIGGERS = ["enter", "leave", "target", "targeted", "death", "kill", "attack", "attacked"] + WORLD_TRIGGERS
UNIT_TRIGGERS = ["friendly-turn", "friendly-turn-end", "enemy-turn", "enemy-turn-end", "spawn"] + TILE_TRIGGERS


class Triggers:

    def __init__(self):
        self._triggers: Dict[str, List[Action]] = {}

    def load(self, parent_id: str, data: dict):
        """ Parse and load a dict defining triggers and it's
        actions into this class

        Args:
            parent_id: The definition id of the unit or tile
            data: The data defining the unit/tile's triggers
        """
        # Iterate through each key in the dict to check it's valid
        for key in data:
            if key == "notes":
                continue

            if not isinstance(data[key], list):
                raise errors.ValidationError(f"Trigger '{key}' must be a list of actions in '{parent_id}'")

            self._triggers[key] = [
                velvet_dawn.mechanics.actions.get_action(parent_id, item)
                for item in data[key]
            ]

    def _run_trigger(self, trigger_name: str, instance: Instance):
        """ Execute each action in a given list of triggers """
        if trigger_name in self._triggers:
            for action in self._triggers[trigger_name]:
                if action.can_run(instance):
                    action.run(instance)

    def on_turn(self, instance: Instance):
        self._run_trigger("turn", instance)

    def on_turn_end(self, instance: Instance):
        self._run_trigger("turn-end", instance)

    def on_friendly_turn(self, instance: Instance):
        self._run_trigger("friendly-turn", instance)

    def on_friendly_turn_end(self, instance: Instance):
        self._run_trigger("friendly-turn-end", instance)

    def on_enemy_turn(self, instance: Instance):
        self._run_trigger("enemy-turn", instance)

    def on_enemy_turn_end(self, instance: Instance):
        self._run_trigger("enemy-turn-end", instance)

    def on_enter(self, instance: Instance):
        self._run_trigger("enter", instance)

    def on_leave(self, instance: Instance):
        self._run_trigger("leave", instance)

    def on_target(self, instance: Instance):
        self._run_trigger("target", instance)

    def on_targeted(self, instance: Instance):
        self._run_trigger("targeted", instance)

    def on_spawn(self, instance: Instance):
        self._run_trigger("spawn", instance)

    def on_game_start(self, instance: Instance):
        self._run_trigger("game", instance)

    def on_death(self, instance: Instance):
        self._run_trigger("death", instance)

    def on_kill(self, instance: Instance):
        self._run_trigger("kill", instance)

    def on_attack(self, instance: Instance):
        self._run_trigger("attack", instance)

    def on_attacked(self, instance: Instance):
        self._run_trigger("attacked", instance)

    def on_round(self, instance: Instance):
        self._run_trigger("round", instance)

File: src/server/src/velvetdawn/core/mechanics/Triggers.java
package velvetdawn.core.mechanics;

import velvetdawn.core.VelvetDawn;
import velvetdawn.core.mechanics.actions.Action;
import velvetdawn.core.mechanics.actions.Actions;
import velvetdawn.core.models.anytype.Any;
import velvetdawn.core.models.anytype.AnyJson;
import velvetdawn.core.models.instances.Instance;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.stream.Collectors;
import java.util.stream.Stream;

public class Triggers {

    /* Triggers class attached to entities and tiles. This
       class also parses the given dict of triggers and
       actions given to the class
     */

    // TODO implement attack and attacked, spawn, death and kill
    // TODO Add validation for this
    Set<String> WORLD_TRIGGERS = Set.of("game", "turn", "turn-end", "round");
    Set<String> TILE_TRIGGERS = Stream.concat(
            Stream.of("enter", "leave", "target", "targeted", "death", "kill", "attack", "attacked"),
            WORLD_TRIGGERS.stream()).collect(Collectors.toSet());
    Set<String> UNIT_TRIGGERS = Stream.concat(
            Stream.of("friendly-turn", "friendly-turn-end", "enemy-turn", "enemy-turn-end", "spawn"),
            TILE_TRIGGERS.stream()).collect(Collectors.toSet());

    public final Map<String, List<Action>> triggers = new HashMap<>();

    /** Parse and load a dict defining triggers and it's
     * actions into this class
     *
     * @param parentId The definition id of the entity or tile
     * @param data The data defining the entity/tile's triggers
     */
    public void load(VelvetDawn velvetDawn, String parentId, AnyJson data) throws Exception {
        // Iterate through each key in the dict to check it 's valid
        for (String key: data.keys()) {
            if (key.equals("notes"))
                continue;

            if (!UNIT_TRIGGERS.contains(key))
                throw new Exception(String.format("Invalid trigger key '%s'", key));

            var errorMessage = String.format("Trigger '%s' must be a list of actions", key);
            var jsonActions = data.get(key, Any.list()).validateInstanceIsList(errorMessage);
            List<Action> actions = new ArrayList<>();
            for (Any action: jsonActions.items) {
                actions.add(Actions.fromJson(velvetDawn, parentId, action.validateInstanceIsJson(errorMessage)));
            }

            this.triggers.put(key, actions);
        }
    }

    /** Execute each action in a given list of triggers */
    private void runTrigger(String triggerName, Instance instance) throws Exception {
        if (this.triggers.containsKey(triggerName)) {
            for (Action action: this.triggers.get(triggerName)) {
                if (action.canRun(instance))
                    action.run(instance);
            }
        }
    }

    public void onTurn(Instance instance) throws Exception {
        this.runTrigger("turn", instance);
    }

    public void onTurnEnd(Instance instance) throws Exception {
        this.runTrigger("turn-end", instance);
    }

    public void onFriendlyTurn(Instance instance) throws Exception {
        this.runTrigger("friendly-turn", instance);
    }

    public void onFriendlyTurnEnd(Instance instance) throws Exception {
        this.runTrigger("friendly-turn-end", instance);
    }

    public void onEnemyTurn(Instance instance) throws Exception {
        this.runTrigger("enemy-turn", instance);
    }

    public void onEnemyTurnEnd(Instance instance) throws Exception {
        this.runTrigger("enemy-turn-end", instance);
    }

    public void onEnter(Instance instance) throws Exception {
        this.runTrigger("enter", instance);
    }

    public void onLeave(Instance instance) throws Exception {
        this.runTrigger("leave", instance);
    }

    public void onTarget(Instance instance) throws Exception {
        this.runTrigger("target", instance);
    }

    public void onTargeted(Instance instance) throws Exception {
        this.runTrigger("targeted", instance);
    }

    public void onSpawn(Instance instance) throws Exception {
        this.runTrigger("spawn", instance);
    }

    public void onGameStart(Instance instance) throws Exception {
        this.runTrigger("game", instance);
    }

    public void onDeath(Instance instance) throws Exception {
        this.runTrigger("death", instance);
    }

    public void onKill(Instance instance) throws Exception {
        this.runTrigger("kill", instance);
    }

    public void onAttack(Instance instance) throws Exception {
        this.runTrigger("attack", instance);
    }

    public void onAttacked(Instance instance) throws Exception {
        this.runTrigger("attacked", instance);
    }

    public void onRound(Instance instance) throws Exception {
        this.runTrigger("round", instance);
    }
}
